joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/callgraph/StaticCallLinker.scala

```diff
@@ -1,64 +1,39 @@
 package io.joern.x2cpg.passes.callgraph
 
 import io.shiftleft.codepropertygraph.Cpg
-import io.shiftleft.codepropertygraph.generated.nodes._
+import io.shiftleft.codepropertygraph.generated.nodes.*
 import io.shiftleft.codepropertygraph.generated.{DispatchTypes, EdgeTypes}
-import io.shiftleft.passes.CpgPass
-import io.shiftleft.semanticcpg.language._
+import io.shiftleft.passes.{ConcurrentWriterCpgPass, ForkJoinParallelCpgPass}
+import io.shiftleft.semanticcpg.language.*
 import org.slf4j.{Logger, LoggerFactory}
 
-import scala.collection.mutable
+class StaticCallLinker(cpg: Cpg) extends ForkJoinParallelCpgPass[Seq[Call]](cpg) {
 
-class StaticCallLinker(cpg: Cpg) extends CpgPass(cpg) {
-
-  import StaticCallLinker._
-  private val methodFullNameToNode = mutable.Map.empty[String, List[Method]]
-
-  override def run(dstGraph: DiffGraphBuilder): Unit = {
+  private val logger: Logger = LoggerFactory.getLogger(classOf[StaticCallLinker])
+  private val BATCH_SIZE     = 100
 
-    cpg.method.foreach { method =>
-      methodFullNameToNode.updateWith(method.fullName) {
-        case Some(l) => Some(method :: l)
-        case None    => Some(List(method))
-      }
-    }
+  override def generateParts(): Array[Seq[Call]] = {
+    cpg.call.grouped(BATCH_SIZE).toArray
+  }
 
-    cpg.call.foreach { call =>
+  override def runOnPart(builder: DiffGraphBuilder, calls: Seq[Call]): Unit = {
+    calls.foreach { call =>
       try {
-        linkCall(call, dstGraph)
+        call.dispatchType match {
+          case DispatchTypes.STATIC_DISPATCH | DispatchTypes.INLINED =>
+            val resolvedMethodOption = cpg.method.fullNameExact(call.methodFullName)
```

Comment:
This is of type `Iterator` not `Option`. Rather place this into memory with `.l` and then you can use this same variable later when you check the size. Right now you're making this query twice when it could be only done once.